### DISCUSSION TITLE
Create contact sheet/grid of images portrait and landscape without gaps

### DISCUSSION THREAD

**Discussion by reaseno** [ImageMagick/ImageMagick#5735] [Help]
I want to create a contact sheet with images from a directory.

I want all images to be displayed above and below each other without gaps. Even with mixed portrait and landscape formats, without cropping the images.

That is, there must be a dynamic height of a row so that the margins are flush and this must be the same for all rows.

Additionally, I want to specify how wide the contact print should be. If you also define the height it could look like this:
![image](https://user-images.githubusercontent.com/3628985/200318464-d47c2371-8f5d-45af-8914-f3cfeac2fc9f.png)

The order of the images must be maintained.

The images normally have the same aspect ratios.

**Comment by snibgo** [comment]
> I want all images to be displayed above and below each other without gaps.

But the example you show has constant thin gaps between images.

> The images normally have the same aspect ratios.

But they might have different aspect ratios? How would that work "without cropping the images"?

The example you show have all images in the aspect ratio 1:sqrt(2), like the standard paper sizes A4, A5 etc, that have the property that chopping one paper sheet in half makes two sheets in the same aspect ratio.

> The order of the images must be maintained.

I don't see how this can be done, or even defined. In your example, what is the order of the images? Are the four small images top-left ordered before the larger image to the right, or perhaps only the first two small images?

Perhaps the "`ashlar:`" format comes close to what you want. It doesn't guarantee that there will be no gaps. But I can't see an algorithm that would guarantee this, with your constraints.

**Reply by reaseno** [reply]
> But the example you show has constant thin gaps between images.

Smaller than any other solution I have seen.

That was just an example and possibly too demanding. But that's how it should be at least:
https://blog.vjeux.com/2012/image/image-layout-algorithm-google-plus.html

**Reply by fmw42** [reply]
your link says on its page:

"The layout algorithm is the consequence of one clever trick: all the images of the same row are have the same height."

Your design does not follow that restriction.

Furthermore; without scale or distorting, you could have an arbitrary set of images that extent beyond the width bounds or do not fill a full row

**Reply by reaseno** [reply]
Yes, this is the point where it can make images bigger or smaller to fill gaps.

My comments were only meant to be examples of the direction it should go.

**Comment by fmw42** [comment]
You would have to manually figure out how to use -append and +append. There is no Imagemagick function that does what you want for an arbitrary set of images with no gaps
